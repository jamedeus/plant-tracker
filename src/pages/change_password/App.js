import React, { useState, useMemo, useRef } from 'react';
import { useTheme } from 'src/context/ThemeContext';
import Navbar from 'src/components/Navbar';
import Cookies from 'js-cookie';
import clsx from 'clsx';

function App() {
    const { ToggleThemeOption } = useTheme();

    const formRef = useRef(null);
    const [oldPassword, setOldPassword] = useState('');
    const [newPassword1, setNewPassword1] = useState('');
    const [newPassword2, setNewPassword2] = useState('');
    const [oldPasswordIncorrect, setOldPasswordIncorrect] = useState(false);
    const [newPasswordError, setNewPasswordError] = useState(false);

    const submit = async (e) => {
        e.preventDefault();
        // Post old and new password to backend
        const response = await fetch("/accounts/change_password/submit/", {
            method: "POST",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
                "X-CSRFToken": Cookies.get('csrftoken')
            },
            credentials: "include",
            body: new URLSearchParams({
                old_password: oldPassword,
                new_password1: newPassword1,
                new_password2: newPassword2,
            }),
        });
        // Redirect to overview if logged in successfully
        if (response.ok) {
            window.location.href = '/';
        // Show correct error message if login failed
        } else {
            const errors = await response.json();
            if (errors.errors.old_password) {
                setOldPasswordIncorrect(true);
            }
            if (errors.errors.new_password2) {
                setNewPasswordError(errors.errors.new_password2[0]);
            }
        }
    };

    const DropdownMenuOptions = useMemo(() => (
        <>
            <li><a onClick={() => window.location.href = "/"}>
                Overview
            </a></li>
            <ToggleThemeOption />
        </>
    ), [ToggleThemeOption]);

    return (
        <div className="container flex flex-col h-screen mx-auto items-center">
            <Navbar
                menuOptions={DropdownMenuOptions}
                title="Change Password"
            />

            <form ref={formRef} className="flex flex-col gap-4 mt-[15vh]">
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Old password</span>
                    </div>
                    <input
                        name="old_password"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            oldPasswordIncorrect && "input-error"
                        )}
                        value={oldPassword}
                        onInput={() => setOldPasswordIncorrect(false)}
                        onChange={(e) => setOldPassword(e.target.value)}
                    />
                </label>
                {oldPasswordIncorrect && (
                    <span className="text-error">
                        Old password incorrect
                    </span>
                )}
                <label className="form-control w-full relative">
                    <div className="label">
                        <span className="label-text">New password</span>
                    </div>
                    <input
                        name="new_password1"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword1}
                        onChange={(e) => setNewPassword1(e.target.value)}
                    />
                </label>
                <label className="form-control w-full">
                    <div className="label">
                        <span className="label-text">Confirm new password</span>
                    </div>
                    <input
                        name="new_password2"
                        type="password"
                        className={clsx(
                            "input w-full input-bordered",
                            newPasswordError && "input-error"
                        )}
                        value={newPassword2}
                        onInput={() => setNewPasswordError(false)}
                        onChange={(e) => setNewPassword2(e.target.value)}
                    />
                </label>
                {newPasswordError && (
                    <span className="text-error">
                        {newPasswordError}
                    </span>
                )}

                <button
                    className="btn btn-success mt-6"
                    onClick={(e) => submit(e)}
                >
                    Change password
                </button>
            </form>
        </div>
    );
}

export default App;
