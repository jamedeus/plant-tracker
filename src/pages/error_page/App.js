import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Navbar from 'src/components/Navbar';
import NavbarDropdownOptions from 'src/components/NavbarDropdownOptions';

function App({ errorMessage }) {
    return (
        <div
            className="container flex flex-col full-screen mx-auto items-center gap-16"
            data-testid="error-page"
        >
            <Navbar
                menuOptions={<NavbarDropdownOptions />}
                title="Error"
                showScanButton={false}
            />

            <div className="flex flex-col text-center text-lg mt-[15vh]">
                {errorMessage}
            </div>

            <Link to="/" className='btn btn-accent' discover="none">
                Go to Overview
            </Link>
        </div>
    );
}

export default App;

App.propTypes = {
    errorMessage: PropTypes.string.isRequired
};
