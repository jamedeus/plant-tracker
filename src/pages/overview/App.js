import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import Navbar from 'src/components/Navbar';
import { useTheme } from 'src/context/ThemeContext';
import GroupCard from 'src/components/GroupCard';
import PlantCard from 'src/components/PlantCard';
import { sendPostRequest, parseDomContext } from 'src/util';
import FilterColumn from 'src/components/FilterColumn';
import FloatingFooter from 'src/components/FloatingFooter';
import Setup from './Setup';
import PrintModal from './PrintModal';
import { useIsBreakpointActive } from "src/useBreakpoint";
import clsx from 'clsx';

const PlantsCol = ({ plants, editing, selectedPlants }) => {
    const openRef = useRef(true);

    return (
        <FilterColumn
            title="Plants"
            contents={plants}
            CardComponent={PlantCard}
            editing={editing}
            selected={selectedPlants}
            openRef={openRef}
            ignoreKeys={[
                'uuid',
                'created',
                'last_watered',
                'last_fertilized',
                'thumbnail'
            ]}
            sortByKeys={[
                {key: 'created', display: 'Added'},
                {key: 'display_name', display: 'Name'},
                {key: 'species', display: 'Species'},
                {key: 'last_watered', display: 'Watered'}
            ]}
            defaultSortKey='created'
            storageKey='overviewPlantsColumn'
        />
    );
};

PlantsCol.propTypes = {
    plants: PropTypes.array.isRequired,
    editing: PropTypes.bool.isRequired,
    selectedPlants: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired
}

const GroupsCol = ({ groups, editing, selectedGroups }) => {
    const openRef = useRef(true);

    return (
        <FilterColumn
            title="Groups"
            contents={groups}
            CardComponent={GroupCard}
            editing={editing}
            selected={selectedGroups}
            openRef={openRef}
            ignoreKeys={[
                'uuid',
                'created'
            ]}
            sortByKeys={[
                {key: 'created', display: 'Added'},
                {key: 'name', display: 'Name'},
                {key: 'location', display: 'Location'}
            ]}
            defaultSortKey='created'
            storageKey='overviewGroupsColumn'
        />
    );
};

GroupsCol.propTypes = {
    groups: PropTypes.array.isRequired,
    editing: PropTypes.bool.isRequired,
    selectedGroups: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired
}

// Render correct components for current state objects
const Layout = ({ plants, groups, selectedPlants, selectedGroups, editing, plantsColRef, groupsColRef, printModalRef }) => {
    // Determines if 2-column layout or single centered column
    const twoColumns = plants.length > 0 && groups.length > 0;

    return (
        <div className={clsx(
            'grid grid-cols-1 mx-auto',
            twoColumns && 'md:grid-cols-2'
        )}>
            {/* Render plants column if 1 or more plants exist */}
            {plants.length > 0 && (
                <div
                    className={clsx(
                        'scroll-mt-20',
                        twoColumns && 'md:mr-12 mb-8 md:mb-0'
                    )}
                    ref={plantsColRef}
                >
                    <PlantsCol
                        plants={plants}
                        editing={editing}
                        selectedPlants={selectedPlants}
                    />
                </div>
            )}
            {/* Render groups column if 1 or more groups exist */}
            {groups.length > 0 && (
                <div
                    className={clsx(
                        'scroll-mt-20',
                        twoColumns && 'md:ml-12'
                    )}
                    ref={groupsColRef}
                >
                    <GroupsCol
                        groups={groups}
                        editing={editing}
                        selectedGroups={selectedGroups}
                    />
                </div>
            )}
            {/* Render setup instructions if database is empty */}
            {plants.length === 0 && groups.length === 0 && (
                <Setup printModalRef={printModalRef} />
            )}
        </div>
    )
};

Layout.propTypes = {
    plants: PropTypes.array.isRequired,
    groups: PropTypes.array.isRequired,
    selectedPlants: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    selectedGroups: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    editing: PropTypes.bool.isRequired,
    plantsColRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    groupsColRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    printModalRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired
}

// Dropdown with links to jump to plant or group columns
// Only rendered on mobile layout (both columns always visible on desktop)
const QuickNavigation = ({ plantsColRef, groupsColRef }) => {
    const jumpToPlants = () => {
        plantsColRef.current.scrollIntoView({
            behavior: "smooth",
            block: "start"
        });
        document.activeElement.blur();
    };

    const jumpToGroups = () => {
        groupsColRef.current.scrollIntoView({
            behavior: "smooth",
            block: "start"
        });
        document.activeElement.blur();
    };

    return (
        <ul
            tabIndex={0}
            className={`menu menu-md dropdown-content mt-3 z-[99] p-2 shadow
                        bg-base-300 rounded-box w-24`}
        >
            <li className="mx-auto"><a onClick={jumpToPlants}>
                Plants
            </a></li>
            <li className="mx-auto"><a onClick={jumpToGroups}>
                Groups
            </a></li>
        </ul>
    );
};

QuickNavigation.propTypes = {
    groupsColRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired,
    printModalRef: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({ current: PropTypes.array }),
    ]).isRequired
}

// Top-left menu button contents
const MenuOptions = ({ archivedOverview, toggleEditing, openPrintModal }) => {
    // Get toggle theme option from context
    const { ToggleThemeOption } = useTheme();

    switch(archivedOverview) {
        case(true):
            return (
                <>
                    <li><a onClick={toggleEditing}>
                        Edit
                    </a></li>
                    <ToggleThemeOption />
                    <li><a href='/'>
                        Main overview
                    </a></li>
                </>
            );
        case(false):
            return (
                <>
                    <li><a onClick={toggleEditing}>
                        Edit
                    </a></li>
                    <li><a onClick={openPrintModal}>
                        Print QR Codes
                    </a></li>
                    <ToggleThemeOption />
                    <li><a href='/archived'>
                        Archived plants
                    </a></li>
                </>
            );
    }
};

MenuOptions.propTypes = {
    archivedOverview: PropTypes.bool.isRequired,
    toggleEditing: PropTypes.func.isRequired,
    openPrintModal: PropTypes.func.isRequired
}

function App() {
    // Load context set by django template
    const [plants, setPlants] = useState(() => {
        return parseDomContext("plants");
    });
    const [groups, setGroups] = useState(() => {
        return parseDomContext("groups");
    });

    // True if desktop layout, false if mobile
    const desktop = useIsBreakpointActive('md');
    // True if mobile layout with stacked plant and group columns
    // False if desktop layout (side by side columns) or only one column
    const stackedColumns = !desktop && plants.length > 0 && groups.length > 0;

    // Check URL to determine if viewing main overview or archive overview
    const archivedOverview = window.location.pathname === '/archived';

    // Request new state from backend if user navigates to overview by pressing
    // back button (last watered/details may be outdated if coming from manage)
    useEffect(() => {
        const handleBackButton = async (event) => {
            if (event.persisted) {
                const response = await fetch('/get_overview_state');
                if (response.ok) {
                    const data = await response.json();
                    setPlants(data['plants']);
                    setGroups(data['groups']);
                } else {
                    alert('Failed to fetch current state, page may be outdated');
                }
            }
        };

        // Add listener on mount, remove on unmount
        if (!archivedOverview) {
            window.addEventListener('pageshow', handleBackButton);
            return () => {
                window.removeEventListener('pageshow', handleBackButton);
            };
        }
    }, []);

    // Create ref for modal used to generate QR codes
    const printModalRef = useRef(null);

    // State object to track edit mode (shows checkbox for each card when true)
    const [editing, setEditing] = useState(false);

    // Track which plant and group checkboxes the user has selected
    const selectedPlants = useRef([]);
    const selectedGroups = useRef([]);

    // Refs used to jump to top of plant and group columns
    const plantsColRef = useRef(null);
    const groupsColRef = useRef(null);

    // Handler for edit option in top-left dropdown
    // Toggle editing state, clear selected, remove focus (closes dropdown)
    const toggleEditing = () => {
        setEditing(!editing);
        selectedPlants.current = [];
        selectedGroups.current = [];
        document.activeElement.blur();
    };

    // Handler for delete button that appears while editing
    const handleDelete = () => {
        // Send delete request for each selected plant, remove uuid from state
        selectedPlants.current.forEach(async plant_id => {
            await sendPostRequest('/delete_plant', {plant_id: plant_id});
        });
        setPlants(plants.filter(
            plant => !selectedPlants.current.includes(plant.uuid))
        );

        // Send delete request for each selected group, remove uuid from state
        selectedGroups.current.forEach(async group_id => {
            await sendPostRequest('/delete_group', {group_id: group_id});
        });
        setGroups(groups.filter(
            group => !selectedGroups.current.includes(group.uuid))
        );

        // Reset editing state
        setEditing(false);
    };

    // Handler for archive button (main overview) and un-archive button
    // (archive overview) that appear while editing. POSTS selected plants and
    // groups to backend then removes from frontend state.
    // Takes bool argument (true if archiving, false if un-archiving)
    const handleArchive = (archived) => {
        // Send archive request for each selected plant, remove uuid from state
        selectedPlants.current.forEach(async plant_id => {
            await sendPostRequest(
                '/archive_plant',
                {plant_id: plant_id, archived: archived}
            );
        });
        setPlants(plants.filter(
            plant => !selectedPlants.current.includes(plant.uuid))
        );

        // Send archive request for each selected group, remove uuid from state
        selectedGroups.current.forEach(async group_id => {
            await sendPostRequest(
                '/archive_group',
                {group_id: group_id, archived: archived}
            );
        });
        setGroups(groups.filter(
            group => !selectedGroups.current.includes(group.uuid))
        );

        // Reset editing state
        setEditing(false);
    };

    return (
        <div className="container flex flex-col min-h-screen mx-auto pb-16">
            <Navbar
                menuOptions={<MenuOptions
                    archivedOverview={archivedOverview}
                    toggleEditing={toggleEditing}
                    openPrintModal={() => printModalRef.current.open()}
                />}
                title={archivedOverview ? "Archived" : "Plant Overview"}
                titleOptions={stackedColumns ? (
                    <QuickNavigation
                        plantsColRef={plantsColRef}
                        groupsColRef={groupsColRef}
                    />
                ) : null}
            />

            <Layout
                plants={plants}
                groups={groups}
                selectedPlants={selectedPlants}
                selectedGroups={selectedGroups}
                editing={editing}
                plantsColRef={plantsColRef}
                groupsColRef={groupsColRef}
                printModalRef={printModalRef}
            />

            <FloatingFooter visible={editing}>
                <button className="btn btn-neutral mr-4" onClick={() => setEditing(false)}>
                    Cancel
                </button>
                {archivedOverview ? (
                    <button className="btn mx-4" onClick={() => handleArchive(false)}>
                        Un-archive
                    </button>
                ) : (
                    <button className="btn mx-4" onClick={() => handleArchive(true)}>
                        Archive
                    </button>
                )}
                <button className="btn btn-error ml-4" onClick={handleDelete}>
                    Delete
                </button>
            </FloatingFooter>

            <PrintModal ref={printModalRef} />
        </div>
    );
}

export default App;
