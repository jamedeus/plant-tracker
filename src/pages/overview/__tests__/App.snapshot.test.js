import createMockContext from 'src/testUtils/createMockContext';
import { PageWrapper } from 'src/index';
import App from '../App';
import { mockContext } from './mockContext';

describe('App', () => {
    beforeAll(() => {
        // Mock page title (set by django template in prod)
        document.title = 'Plant Overview';
    });

    it('matches snapshot when plants and groups exist (desktop layout)', () => {
        // Create mock state objects with a single plant and group
        createMockContext('plants', [mockContext.plants[0]]);
        createMockContext('groups', [mockContext.groups[0]]);
        createMockContext('show_archive', mockContext.show_archive);
        createMockContext('user_accounts_enabled', true);

        // Set width greater than tailwind md breakpoint
        window.innerWidth = 800;

        // Render App, confirm matches snapshot
        const component = render(
            <PageWrapper>
                <App />
            </PageWrapper>
        );
        expect(component).toMatchSnapshot();
    });

    it('matches snapshot when plants and groups exist (mobile layout)', () => {
        // Create mock state objects with a single plant and group
        createMockContext('plants', [mockContext.plants[0]]);
        createMockContext('groups', [mockContext.groups[0]]);
        createMockContext('show_archive', mockContext.show_archive);
        createMockContext('user_accounts_enabled', true);

        // Set width less than tailwind md breakpoint
        window.innerWidth = 600;

        // Render App, confirm matches snapshot
        const component = render(
            <PageWrapper>
                <App />
            </PageWrapper>
        );
        expect(component).toMatchSnapshot();
    });

    it('matches snapshot when only plants exist', () => {
        // Create mock state objects with a single plant
        createMockContext('plants', [mockContext.plants[0]]);
        createMockContext('groups', []);
        createMockContext('show_archive', mockContext.show_archive);
        createMockContext('user_accounts_enabled', true);

        // Render App, confirm matches snapshot
        const component = render(
            <PageWrapper>
                <App />
            </PageWrapper>
        );
        expect(component).toMatchSnapshot();
    });

    it('matches snapshot when only groups exist', () => {
        // Create mock state objects with a single group
        createMockContext('plants', []);
        createMockContext('groups', [mockContext.groups[0]]);
        createMockContext('show_archive', mockContext.show_archive);
        createMockContext('user_accounts_enabled', true);

        // Render App, confirm matches snapshot
        const component = render(
            <PageWrapper>
                <App />
            </PageWrapper>
        );
        expect(component).toMatchSnapshot();
    });

    it('matches snapshot when no models exist (setup)', () => {
        // Create mock state objects
        createMockContext('plants', []);
        createMockContext('groups', []);
        createMockContext('show_archive', false);
        createMockContext('user_accounts_enabled', true);

        // Render App, confirm matches snapshot
        const component = render(
            <PageWrapper>
                <App />
            </PageWrapper>
        );
        expect(component).toMatchSnapshot();
    });
});

describe('App (archived page)', () => {
    beforeAll(() => {
        // Mock page title (set by django template in prod)
        document.title = 'Archived';
    });

    it('matches snapshot when plants and groups exist', () => {
        // Create mock state objects with a single plant and group (flip
        // archived bools to true)
        createMockContext('plants', [{
            ...mockContext.plants[0],
            archived: true
        }]);
        createMockContext('groups', [{
            ...mockContext.groups[0],
            archived: true
        }]);
        createMockContext('user_accounts_enabled', true);

        // Mock window.location to simulate archived overview
        Object.defineProperty(window, 'location', {
            configurable: true,
            value: {
                ...window.location,
                href: 'https://plants.lan/',
                pathname: '/archived',
                assign: jest.fn()
            }
        });

        // Render App, confirm matches snapshot
        const component = render(
            <PageWrapper>
                <App />
            </PageWrapper>
        );
        expect(component).toMatchSnapshot();
    });
});

