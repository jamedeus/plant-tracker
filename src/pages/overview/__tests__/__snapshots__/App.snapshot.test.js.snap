// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot when plants and trays exist 1`] = `
[
  <div
    className="container flex flex-col min-h-screen mx-auto pb-16"
  >
    <div
      className="navbar bg-base-100 mb-4 sticky top-0 z-[99]"
    >
      <div
        className="navbar-start w-auto"
      >
        <div
          className="dropdown"
        >
          <div
            className="btn btn-ghost btn-circle"
            role="button"
            tabIndex={0}
          >
            <svg
              className="h-5 w-5"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h7"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
              />
            </svg>
          </div>
          <ul
            className="menu menu-md dropdown-content mt-3 z-[99]
                                    p-2 shadow bg-base-300 rounded-box w-52"
            tabIndex={0}
          >
            <li>
              <a
                onClick={[Function]}
              >
                Edit
              </a>
            </li>
            <li>
              <a
                onClick={[Function]}
              >
                Print QR Codes
              </a>
            </li>
            <li>
              <a
                onClick={[Function]}
              >
                Light mode
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        className="navbar-center mx-auto"
      >
        <div
          className="dropdown dropdown-center"
        >
          <a
            className="btn btn-ghost"
            role="button"
            style={
              {
                "fontSize": "32px",
              }
            }
            tabIndex={0}
          >
            Plant Overview
          </a>
          <div
            className="dropdown-content z-[1] flex"
            tabIndex={0}
          />
        </div>
      </div>
      <div
        className="navbar-end w-auto invisible"
      >
        <div
          className="btn btn-ghost btn-circle"
          role="button"
          tabIndex={0}
        >
          <svg
            className="h-5 w-5"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h7"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="grid grid-cols-1 md:grid-cols-2 mx-auto"
    >
      <div
        className="md:mr-12 mb-8 md:mb-0"
      >
        <div
          className="collapse bg-base-200 w-96 px-4 mx-auto max-w-90 md:max-w-full"
        >
          <input
            defaultChecked={true}
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="collapse-title text-xl font-medium text-center"
          >
            Plants (1)
          </div>
          <div
            className="collapse-content min-w-0"
          >
            <div
              className="px-4 mb-4 relative"
            >
              <input
                className="input input-bordered w-full text-center pr-10"
                onChange={[Function]}
                placeholder="filter"
                style={
                  {
                    "textIndent": "1.625rem",
                  }
                }
                type="text"
                value=""
              />
              <button
                className="btn-close absolute right-6 top-2
                        opacity-0 -z-10"
                onClick={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="w-8 h-8"
                  data-slot="icon"
                  fill="currentColor"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.28 4.22a.75.75 0 0 0-1.06 1.06L6.94 8l-2.72 2.72a.75.75 0 1 0 1.06 1.06L8 9.06l2.72 2.72a.75.75 0 1 0 1.06-1.06L9.06 8l2.72-2.72a.75.75 0 0 0-1.06-1.06L8 6.94 5.28 4.22Z"
                  />
                </svg>
              </button>
            </div>
            <div
              className="mb-4"
            >
              <div
                className="card card-side bg-neutral text-neutral-content mx-auto relative
                     cursor-pointer
                "
                onClick={[Function]}
              >
                <figure
                  className="h-24 w-20 min-h-20 min-w-20"
                >
                  <img
                    className="w-full h-full object-cover"
                    src="/media/thumbnails/photo_thumb.jpg"
                  />
                </figure>
                <div
                  className="card-body my-auto text-start"
                >
                  <h2
                    className="card-title line-clamp-1 pr-8"
                  >
                    Test Plant
                  </h2>
                  <span>
                    <svg
                      aria-hidden="true"
                      className="svg-inline--fa fa-droplet mr-2 text-info"
                      data-icon="droplet"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      style={{}}
                      viewBox="0 0 384 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M192 512C86 512 0 426 0 320C0 228.8 130.2 57.7 166.6 11.7C172.6 4.2 181.5 0 191.1 0h1.8c9.6 0 18.5 4.2 24.5 11.7C253.8 57.7 384 228.8 384 320c0 106-86 192-192 192zM96 336c0-8.8-7.2-16-16-16s-16 7.2-16 16c0 61.9 50.1 112 112 112c8.8 0 16-7.2 16-16s-7.2-16-16-16c-44.2 0-80-35.8-80-80z"
                        fill="currentColor"
                        style={{}}
                      />
                    </svg>
                    4 days ago
                  </span>
                </div>
                <button
                  className="btn-close absolute right-2 top-8 z-40"
                  onClick={[Function]}
                  role="button"
                  tabIndex={0}
                >
                  <svg
                    aria-hidden="true"
                    className="w-8 h-8"
                    data-slot="icon"
                    fill="currentColor"
                    viewBox="0 0 16 16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clipRule="evenodd"
                      d="M4.22 6.22a.75.75 0 0 1 1.06 0L8 8.94l2.72-2.72a.75.75 0 1 1 1.06 1.06l-3.25 3.25a.75.75 0 0 1-1.06 0L4.22 7.28a.75.75 0 0 1 0-1.06Z"
                      fillRule="evenodd"
                    />
                  </svg>
                </button>
              </div>
              <div
                className="collapse bg-neutral rounded-t-none
                
            "
              >
                <input
                  className="hidden"
                  defaultChecked={false}
                  onChange={[Function]}
                  type="checkbox"
                />
                <div
                  className="collapse-content"
                >
                  <p
                    className="flex"
                  >
                    <span
                      className="font-semibold"
                    >
                      Species:
                    </span>
                    <span
                      className="ml-auto text-right"
                    >
                      Calathea
                    </span>
                  </p>
                  <p
                    className="flex"
                  >
                    <span
                      className="font-semibold"
                    >
                      Pot size:
                    </span>
                    <span
                      className="ml-auto text-right"
                    >
                      4
                    </span>
                  </p>
                  <div
                    className="hidden"
                  >
                    <p
                      className="font-semibold mt-3"
                    >
                      Description:
                    </p>
                    <p
                      className="text-sm line-clamp-6"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="md:ml-12"
      >
        <div
          className="collapse bg-base-200 w-96 px-4 mx-auto max-w-90 md:max-w-full"
        >
          <input
            defaultChecked={true}
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="collapse-title text-xl font-medium text-center"
          >
            Trays (1)
          </div>
          <div
            className="collapse-content min-w-0"
          >
            <div
              className="px-4 mb-4 relative"
            >
              <input
                className="input input-bordered w-full text-center pr-10"
                onChange={[Function]}
                placeholder="filter"
                style={
                  {
                    "textIndent": "1.625rem",
                  }
                }
                type="text"
                value=""
              />
              <button
                className="btn-close absolute right-6 top-2
                        opacity-0 -z-10"
                onClick={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="w-8 h-8"
                  data-slot="icon"
                  fill="currentColor"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.28 4.22a.75.75 0 0 0-1.06 1.06L6.94 8l-2.72 2.72a.75.75 0 1 0 1.06 1.06L8 9.06l2.72 2.72a.75.75 0 1 0 1.06-1.06L9.06 8l2.72-2.72a.75.75 0 0 0-1.06-1.06L8 6.94 5.28 4.22Z"
                  />
                </svg>
              </button>
            </div>
            <div
              className="mb-4"
            >
              <div
                className="card bg-neutral text-neutral-content mx-auto w-full cursor-pointer"
                onClick={[Function]}
              >
                <div
                  className="card-body text-center"
                >
                  <h2
                    className="card-title mx-auto"
                  >
                    Test tray
                  </h2>
                  <p>
                    Contains 
                    4
                     plants
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <dialog
      className="modal"
    >
      <div
        className="modal-box text-center flex flex-col pt-4 "
      >
        <form
          method="dialog"
        >
          <button
            className="btn-close absolute right-4 top-4"
          >
            <svg
              aria-hidden="true"
              className="w-8 h-8"
              data-slot="icon"
              fill="currentColor"
              viewBox="0 0 16 16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.28 4.22a.75.75 0 0 0-1.06 1.06L6.94 8l-2.72 2.72a.75.75 0 1 0 1.06 1.06L8 9.06l2.72 2.72a.75.75 0 1 0 1.06-1.06L9.06 8l2.72-2.72a.75.75 0 0 0-1.06-1.06L8 6.94 5.28 4.22Z"
              />
            </svg>
          </button>
        </form>
        <h3
          className="font-bold text-lg mb-6"
        >
          Select QR Code Size
        </h3>
        <div
          className="h-36 flex flex-col justify-center"
        >
          <div
            aria-orientation="horizontal"
            className="tab-group"
            role="tablist"
          >
            <button
              aria-selected={true}
              className="tab-option tab-option-selected"
              data-headlessui-state="selected"
              id="headlessui-tabs-tab-:r0:"
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="tab"
              tabIndex={0}
              type="button"
            >
              small
            </button>
            <button
              aria-selected={false}
              className="tab-option "
              data-headlessui-state=""
              id="headlessui-tabs-tab-:r1:"
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="tab"
              tabIndex={-1}
              type="button"
            >
              medium
            </button>
            <button
              aria-selected={false}
              className="tab-option "
              data-headlessui-state=""
              id="headlessui-tabs-tab-:r2:"
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="tab"
              tabIndex={-1}
              type="button"
            >
              large
            </button>
          </div>
          <div
            className="mt-4"
          >
            <div
              data-headlessui-state="selected"
              id="headlessui-tabs-panel-:r3:"
              role="tabpanel"
              tabIndex={0}
            >
              <span
                className="text-center"
              >
                About ¾ inch wide
                <br />
                96
                 QR codes per sheet
              </span>
            </div>
            <span
              aria-hidden="true"
              id="headlessui-tabs-panel-:r4:"
              role="tabpanel"
              style={
                {
                  "borderWidth": "0",
                  "clip": "rect(0, 0, 0, 0)",
                  "height": 0,
                  "left": 1,
                  "margin": -1,
                  "overflow": "hidden",
                  "padding": 0,
                  "position": "fixed",
                  "top": 1,
                  "whiteSpace": "nowrap",
                  "width": 1,
                }
              }
              tabIndex={-1}
            />
            <span
              aria-hidden="true"
              id="headlessui-tabs-panel-:r5:"
              role="tabpanel"
              style={
                {
                  "borderWidth": "0",
                  "clip": "rect(0, 0, 0, 0)",
                  "height": 0,
                  "left": 1,
                  "margin": -1,
                  "overflow": "hidden",
                  "padding": 0,
                  "position": "fixed",
                  "top": 1,
                  "whiteSpace": "nowrap",
                  "width": 1,
                }
              }
              tabIndex={-1}
            />
          </div>
        </div>
        <div
          className="modal-action mx-auto"
        >
          <button
            className="btn btn-accent"
            onClick={[Function]}
          >
            Generate
          </button>
        </div>
      </div>
      <form
        className="modal-backdrop"
        method="dialog"
      >
        <button>
          close
        </button>
      </form>
    </dialog>
  </div>,
  <dialog
    className="modal"
  >
    <div
      className="modal-box text-center flex flex-col pt-4 "
    >
      <form
        method="dialog"
      >
        <button
          className="btn-close absolute right-4 top-4"
        >
          <svg
            aria-hidden="true"
            className="w-8 h-8"
            data-slot="icon"
            fill="currentColor"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.28 4.22a.75.75 0 0 0-1.06 1.06L6.94 8l-2.72 2.72a.75.75 0 1 0 1.06 1.06L8 9.06l2.72 2.72a.75.75 0 1 0 1.06-1.06L9.06 8l2.72-2.72a.75.75 0 0 0-1.06-1.06L8 6.94 5.28 4.22Z"
            />
          </svg>
        </button>
      </form>
      <h3
        className="font-bold text-lg mb-6"
      >
        Error
      </h3>
      <div
        className="min-h-36 flex flex-col justify-center mx-auto"
      />
      <div
        className="modal-action mx-auto"
      >
        <button
          className="btn"
          onClick={[Function]}
        >
          OK
        </button>
      </div>
    </div>
    <form
      className="modal-backdrop"
      method="dialog"
    >
      <button>
        close
      </button>
    </form>
  </dialog>,
]
`;
